### DISCUSSION TITLE
Custom Bounds based on concave polygon

### DISCUSSION THREAD

**Discussion by NordelDev** [dyn4j/dyn4j#210] [Q&A]
Hi William,

I am currently working on an implementation of a custom `Bounds` class, which is based on a `Polygon`, which might be concave.

Since I need the functionality of the `BoundsListener` I also need to implement the required detection method if `SimulationBody`s are inside the `Bounds`.

My idea is the following:

- Define the custom `Bounds` per `Polygon` definition
- Decompose the `Polygon` in case it is concave
- Do some initial checks, if a `SimulationBody` should be checked by the `BoundsListener`
- Calculate the overlapping area of each `Shape` of interest of the `BodyFixture`s with the decomposed `Shape`s of the custom `Bounds`
- When the total sum of overlapping area is 0, then the `SimulationBody` is outside the `Bounds`

![grafik](https://user-images.githubusercontent.com/34217758/147381245-d2120bb0-c44b-45ca-afeb-0120262acbbc.png)

**Question:**

Are there already some built-in functionalities, which allows the calculation of the overlapping area of 'Convex' 'Shape's? So far I was not able to find something but maybe the methods are in a class which I didn't check.

If there are no functionallities implemented do you plan to implement them?

---

Another use-case I could require them, is to check if `SimulationBody`s are completly within a certain sensor area of another `SimulationBody`.

![grafik](https://user-images.githubusercontent.com/34217758/147381339-e31e5b79-9745-482d-94b8-9243bf5bb70c.png)

---

Or do you have other suggestions how I could achieve my desired goals without implementing my own 'Shape' class modifications?

Any idea is much appreciated :)

**Comment by wnbittle** [comment]
No, there aren't any methods for computing the intersection area of two convex shapes (generalizing a bit).  The area of two overlapping convex `Polygon`s boils down to computing the overlapping polygon, and calculating the area of that intersection.  This could get expensive as the number of bodies in the simulation grow.  Whether something like this would be added?  I'm open to adding something like this and I found a few algorithms that are reasonably simple to implement.  Note, dyn4j supports non-polygonal shapes as well: `Circle`, `Capsule`, `Ellipse`, `Slice`, etc. and these would not be supported.

You shouldn't need the area though, but I can see that `BoundsListener` interface is a bit rigid for what you are trying to do.  Right now, the `World` class (and it's super classes) ONLY call the `BoundsListener.isOutside(AABB)` method which means there isn't a good way to extend it with custom functionality - particularly what you are trying to do.  If this was changed to call a new method with `BoundsListener.isOutside(AABB, Transform, Convex)` signature, for example, then you'd be able to do what you need to by extending the `Bounds` interface.  You could use the AABB to do a quick check for intersection, then use the Convex to do a more accurate intersection test via `Sat` or `Gjk`.  Technically, this could be done without the `Bounds` / `BoundsListener` functionality via sensors, but I understand that the bounds functionality is convenient.

```java
boolean isOutside(AABB aabb, Transform tx, Convex convex) {
  // check the AABBs first
  if (!aabb.overlaps(this.aabb)) {
    return true;
  }

  // check the actual shapes now
  if (!gjk.detect(this.transform, this.convex, tx, convex)) {
    // then this means the AABBs are overlapping, but the convexes are not
    return true;
  }

  return false;
}
```

The `World` class (and it's super classes) already accumulate the results from all the isOutside calls and calls the `BoundsListener` methods if all `Fixtures` are outside the bounds.

The sensor scenario is the opposite and harder - detecting whether something is completely inside vs. completely outside.  In the case of a `Circle` sensor and `Polygon` other shapes (which was probably just an example) it's easy.  If the sensor detects a collision, then check if the colliding shape's vertices are all within the radius of the circle from the circle center.  Another alternative would be to implement the SAT algorithm which would be able to tell if there is full containment.  The SAT algorithm projects the shapes along a number line, for every axis of the `Polygon`, producing two intervals where you can check for containment.  If all axis projects have containment, then the shape is fully contained within the other.  The `Sat` class detects this, but uses the information for a different purpose.  Maybe this could give you some ideas.  Another thought I had was if I created a new interface called `ContainmentDetector` that `Sat` and `Gjk` could implement - I haven't fully thought it through with `Gjk`, but `Sat` for sure could support this.

I'll add all these as enhancement issues.  Let me know if you think these would work for you and I can try to find some time to get them implemented.

**Reply by NordelDev** [reply]
Thank you very much for your detailed answer and your good suggestions.

If you are interessted in implementing an updated `BoundsListener` and some intersecting methods I really would appreciate that. I don't like to make my own adjustments to the core library to avoid possible problems in a new release. :-D

I will have a deeper look at `SAT` and `GJK`, which seems like would be sufficient enough for my goal with the sensor in realtion to a `CollisionListener`. 

Again thank you for your reply

**Comment by wnbittle** [comment]
@NordelDev Review 4.2.1 which should contain the enhancements that I discussed above - any additional testing would be appreciated.  I would just caution that you read the javadoc documentation on the new features very carefully since some decisions were made in degenerate cases that might impact your use case slightly.

**Reply by NordelDev** [reply]
You are awesome! Thank you very much for your quick actions and that you heed my questions and even implement them as features.

I will test the new functionalities with my intended use case and give you feedback. (I leave this thread/discussion as not answered to post my feedback later)
